### ISSUE TITLE
Questions on Jekyll and MathJax versions

### ISSUE THREAD

**Issue by netw0rkf10w** [fmeum/jekyll-mathjax-csp#8]
Hi,

First of all thanks for the software!

I am considering switching from Hugo just because of this! Before proceeding, I would like to ask some questions:

1. Does your plugin work with the latest version of Jekyll?

2. It seems that you use the deprecated `mathjax-node` (which was a workaround for MathJax 2.x server-side rendering). Since MathJax 3.0, server-side rendering is natively supported (please see [here](https://github.com/mathjax/MathJax-demos-node) for example). Do you plan to upgrade to MathJax 3?

Thank you very much in advance!

**Comment by fmeum** [comment]
Thank you very much for your interest in this project!

> 1. Does your plugin work with the latest version of Jekyll?

Yes, `jekyll-mathjax-csp` is compatible with Jekyll 4.x (just confirmed with 4.1.1).

> 2. It seems that you use the deprecated `mathjax-node` (which was a workaround for MathJax 2.x server-side rendering). Since MathJax 3.0, server-side rendering is natively supported (please see [here](https://github.com/mathjax/MathJax-demos-node) for example). Do you plan to upgrade to MathJax 3?

Thanks for making me aware of the native support for server-side rendering. I will see whether I can get it to work and would then release a new major version of `jekyll-mathjax-csp`.

**Comment by fmeum** [comment]
I just tried out the `tex2svg-page` tool that is available at the link you provided and found it to be a suitable replacement for `mathjax-node`. I will need to make some architectural changes, but this may even end up simplifying the plugin setup. 

I will keep you updated on my progress via this issue.

**Comment by netw0rkf10w** [comment]
That's really awesome! Thank you for the great support! I look forward to hearing good news soon.

**Comment by fmeum** [comment]
I switched to MathJax 3 on the [switch_to_mathjax_3 branch](https://github.com/FabianHenneke/jekyll-mathjax-csp/tree/switch_to_mathjax_3).

If you are using Bundler, you can try out the new version by adding the following line to your `Gemfile`:

```ruby
gem "jekyll-mathjax-csp", github: "FabianHenneke/jekyll-mathjax-csp", branch: "switch_to_mathjax_3"
```

and following the setup instructions of the [README](https://github.com/FabianHenneke/jekyll-mathjax-csp/tree/switch_to_mathjax_3) (apart from step 2).

Please let me know how that went for you.

**Comment by fmeum** [comment]
@MaximeKjaer You might also be interested in this thread/change.

**Comment by netw0rkf10w** [comment]
@FabianHenneke Thank you very much! Let me give it a try.

**Comment by netw0rkf10w** [comment]
Hi @FabianHenneke,

I am not quite sure to understand the 4th step in your instructions:

> 4. Add the {% mathjax_csp_sources %} Liquid tag where you want the CSP 'sha256-...' hashes for <style> elements to be emitted. Don't forget to add the YAML front matter (two lines of ---) to such files. If you specify your CSP in a different way, add the style-src sources the plugins prints to the console during build.

Could you please be more specific? I followed the first 3 steps but the maths are not displayed (I'm testing locally), so I guess this final step is crucial.

Thank you in advance!

**Comment by MaximeKjaer** [comment]
Thank you for notifying me of this, @FabianHenneke ! I just had a look at the implementation and tried it out, and the change on `switch_to_mathjax_3` looks great. Overall, build time of my full site is down from ~80 seconds (when building to SVG) to ~26 seconds by switching to this -- that's amazing!

However, migrating broke all of my inline math, as the `single_dollar` option isn't supported on that branch right now. From a quick look at the MathJax docs, it looks like we could pass something like `inlineMath: [['$', '$']]` onto the `Tex` constructor if the option has been set. Likewise for the CommonHTML output, which I rely on (my use-case is that I have a lot of inline math, and having inline SVG breaks copy-pasting paragraphs).

**Comment by MaximeKjaer** [comment]
@netw0rkf10w The 4th step basically describes a file that looks like this:

```
---
---
// Some code here
{% mathjax_csp_sources %}
// more code here
```

The `---` lines are there just to make sure that Jekyll processes the `{% mathjax_csp_sources %}` tag as Liquid, and not as plain text (if I remember correctly, I haven't written any Jekyll code in a while!). 

This step shouldn't be crucial to get math to render; you should only follow it if you want to add CSP to your site, but that shouldn't be needed when you're just building locally for now. 

To get the math to render, perhaps make sure that you are using `$$` as your math delimiter, like so:

```
$$a^2 + b^2 = c^2$$
```

If that still doesn't work, you can post a snippet of the code for the math you're rendering, and what HTML is being rendered on the site, and I can take a look to see if there's anything wrong :)

**Comment by netw0rkf10w** [comment]
@MaximeKjaer That's very kind of you. Thanks a lot!

As a test, a few days ago I switched to the classical client-side MathJax to see if it worked, and it did. Today I switched back to `jekyll-mathjax-csp` in order to have more information to share with you, but surprisingly it's working this time. I really don't know what happened.

> However, migrating broke all of my inline math, as the `single_dollar` option isn't supported on that branch right now. From a quick look at the MathJax docs, it looks like we could pass something like `inlineMath: [['$', '$']]` onto the `Tex` constructor if the option has been set. Likewise for the CommonHTML output, which I rely on (my use-case is that I have a lot of inline math, and having inline SVG breaks copy-pasting paragraphs).

I guess we can change [this line](https://github.com/FabianHenneke/jekyll-mathjax-csp/blob/133f02b043fb6ff997e98bd0d9f641da7924239b/bin/mathjaxify#L97) in the code:

`const tex = new TeX({packages: AllPackages});`

to

`const tex = new TeX({packages: AllPackages, inlineMath: [['$', '$'], ['\\(', '\\)']]});`

I will give it a try to see if it works. Ideally it would be great to have an option to parse a custom MathJax configuration to `jekyll-mathjax-csp`.

**Comment by netw0rkf10w** [comment]
> `const tex = new TeX({packages: AllPackages, inlineMath: [['$', '$'], ['\\(', '\\)']]});`

This works like a charm! Now both double-dollars and single-dollars work for me. But is it normal that they have exactly the same behavior? I would expect the double-dollars to center the equation on a new line, as this is standard in TeX...

**Comment by netw0rkf10w** [comment]
Everything is working fine on my side, except maybe a final issue:

Unnumbered equations produced by the `equation*` (or `align*`) environment are not displayed. For example, this equation

```latex
\begin{equation*} a + b = c \end{equation*}
```

is displayed as:

> \begin{equation _} a + b = c \end{equation_ }

(Everything between the two stars are italicized). @FabianHenneke @MaximeKjaer Would you have an idea of how to fix this please?

**Comment by fmeum** [comment]
I'm somewhat busy right now and won't be able to look into this right now.

Maxime is right in that step 4 of the setup instructions only matters if you care about Content Security Policies. If you have to google that, you most likely don't, so don't worry about that.

@MaximeKjaer @netw0rkf10w If either you figures out how to fix this and sends a PR to the `switch_to_mathjax_3` branch, I'll review it. Otherwise I will see whether I can take a look on the weekend.

**Comment by MaximeKjaer** [comment]
@netw0rkf10w Seems like kramdown is parsing the asterisks as start and end of Markdown italics. I think this stems from the fact that this plugin runs after kramdown parsing; I often have a similar problem in inline math with underscores. You can try to escape the asterisks, like so: `\*` (see the [emphasis section of the kramdown docs](https://kramdown.gettalong.org/syntax.html#emphasis)).

**Comment by netw0rkf10w** [comment]
@MaximeKjaer Thanks. I'll use `\*` as a workaround.

Thanks also for the PR! I don't know why you obtained blurry output, but it looks perfect on my side...

**Comment by satyanash** [comment]
Saw a 50% reduction in regeneration speed from ~12 seconds to ~6sec after switching to the `switch_to_mathjax_3` branch :rocket:

**Comment by satyanash** [comment]
For now, I have implemented batch mode on top of MathJax 3 (`switch_to_mathjax_3`) only https://github.com/fmeum/jekyll-mathjax-csp/pull/10

I think we should be able to backport this to `master` as well.

**Comment by fmeum** [comment]
@satyanash Thanks for the contribution! I'm not using this plugin myself anymore, but I will review your PR and publish a new major version based on the `switch_to_mathjax_3` branch.

**Comment by fmeum** [comment]
@satyanash @MaximeKjaer @netw0rkf10w I have rebased and merged `switch_to_mathjax_3` into `master`. Could you verify that `master` works for you? I won't be able to test it myself, but once you confirm it works, I will publish a new gem version.

**Comment by satyanash** [comment]
> I will review your PR and publish a new major version based on the switch_to_mathjax_3 branch.

Hi @fmeum, thanks for picking this up! :slightly_smiling_face: 

> Could you verify that master works for you?

For me the new `master` is working fine! The math renders as it did before on the branch, before the rebase.

**Comment by fmeum** [comment]
I just released version 2.0.0.
